README.rst
============
``gs.image``
============
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Image support for GroupServer
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Authors: `Michael JasonSmith`_,
         Richard Waid
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2014-11-14
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.net`_.

..  _Creative Commons Attribution-Share Alike 4.0 International License:
    http://creativecommons.org/licenses/by-sa/4.0/

Introduction
============

GroupServer_ displays images in many places, including on
profiles [#profiles]_, in posts [#posts]_, and on the Image page
[#image]_. This egg wraps some functions provided by PIL_ to
provide support to the other products for resizing and caching
images. The functionality of this product is primarily provided
by two classes:

* GSImage_ is used to resize and implicitly cache images. 
* The GSSquareImage_ class resizes and crops the images to ensure
  they are square, before implicitly caching the result.

The resizing is done by one of three utilities_.

``GSImage``
===========

The ``gs.image.GSImage`` class provides an application cache for
images. This means the computationally-expensive task of creating
a thumbnail only has to be done once. It also provides convenient
methods_ for integrating and resizing the image.

Constructor
-----------

::

 GSImage(data)

``data``:
  Either a ``file`` or a byte-array (string) containing the
  image.

Properties
----------

``data``:
  The image data.

``width``:
  The width of the image.

``height``:
  The height of the image.

``contentType``:
  The content type of the image (useful for setting the HTTP
  response).

Methods
-------

``getSize()``:
  Returns the size of the image in bytes.

``getImageSize()``:
  Returns a 2-tuple of ``(GSImage.width, GSImage.height)``.

``get_resized(x, y, maintain_aspect=True, only_smaller=True, return_cache_path=False)``:
  Returns a new ``GSImage``, with a maximum width of ``x`` and a
  maximum height of ``y``.

  * If ``maintain_aspect`` is ``True`` then the aspect ratio of
    the new image (``x/y``) will be the same as the aspect ratio
    of the old image.
  * If ``only_smaller`` is ``True`` then the image will be scaled
    only if it is smaller.
  * Finally [#command]_, if ``return_cache`` is ``True`` then the
    path to the file in the cache is returned, not the image (see
    `Side Effects`_ below).

Side Effects
------------

The main job of the ``GSImage`` class is to *implicitly* cache
the scaled image when ``get_resized`` is called. The cache is
located at
``{clientHome}/groupserver.data/groupserver.GSImage.cache``,
where ``clientHome`` is the directory of the GroupServer instance
(normally ``var/instance`` in the installation directory).

The file-name is made up of ``{md5}{width}x{height}x{maintainAspect}``:

* ``md5``: An MD5 Sum of the image data.
* ``width``: The width of the image.
* ``height``: The height of the image.
* ``maintainAspect``: If the aspect ratio was kept.

The cached image will be of the same type as the original image. However,
the quality_ may be set to a different value.

Quality
~~~~~~~

If the base image is a JPEG, then the quality of the image is set
depending on the size of the *smallest* dimension. The idea is
that the low quality is less noticeable at small sizes. For large
sizes progressive-rendering is turned on, so the time-to-glass is
reduced. The different quality settings at different sizes is as
follows:

================ =======
Size             Quality
================ =======
s <= 50px             40
50 < s <= 200px       60
s > 200px             75  
================ =======

If the base image is something other than a JPEG (such as a PNG)
then it is simply resized.

Examples
--------

Resize the image in ``fileName``, ensuring neither dimension is
more than 128px::

  f = file(fileName)
  i = GSImage(f)
  scaledImage = i.get_resized(128, 128)

Return a scaled image from a Zope page view by over-writing the
``__call__`` method::

    class MyImage(Products.Five.BrowserView):
        def __init__(self, context, request):
            super(MyImage, self).__init__(context, request)

        ...
    
        def __call__(self):
            f = file(self.fileName)
            i = GSImage(f)
            scaledImage = i.get_resized(128, 128)

            self.request.RESPONSE.setHeader('Content-Type',
                                            scaledImage.contentType)
            self.request.RESPONSE.setHeader('Content-Length',
                                            scaledImage.getSize())
            retval = scaledImage.data
            return retval


``GSSquareImage``
=================

The ``gs.image.GSSquareImage`` class resizes images, just like
the parent GSImage_ class, but all images are made square. It
inherits the same constructor_, and properties_ as its
parent. However, the `get_resized`_ method is different.

``get_resized``
---------------

Get a resized square image.

:Synopsis: ``get_resized(size)``

:Description: The ``get_resized`` method resizes the image, so
              neither the width nor the height will exceed the
              ``size``, and both will be the same. See
              `thumbnail_img_square`_ for more details on the
              algorithm used to do this.

:Arguments: ``size`` the maximum width and height of the image in
            pixels.

:Returns: A new ``GSSquareImage``.

Example
-------

Create a square image, 32 pixels on a side::

  f = file(fileName)
  i = GSSquareImage(f)
  scaledImage = i.get_resized(32)


Utilities
=========

Three utilities are provided 

#. `thumbnail_img`_ 
#. `thumbnail_img_noaspect`_
#. `thumbnail_img_square`_

``thumbnail_img``
-----------------

Create a thumbnail of an image.

:Synopsis: ``thumbnail_img(i, x, y, method=Image.ANTIALIAS)``

:Description: The ``thumbnail_img`` utility creates a new image, scaled
          down so the width will not exceed ``x``, and height will not
          exceed ``y``, and the aspect-ratio (``x/y``) will be
          maintained. It is similar to ``Image.thumbnail``, but it returns
          a new image, rather than working on the image in-place.

:Arguments: 
 
  ``i``:
    The PIL_ image to resize.

  ``x``: 
    The maximum width.

  ``y``:
    The maximum height.

  ``method``
    The scaling method.

:Returns: A new image. The width will not exceed ``x``, and height will not
          exceed ``y``, and the aspect-ratio (``x/y``) will be maintained.

Example
~~~~~~~

Scale a PIL image so neither dimension is more that 127px::

   i = PIL.Image.open(data)
   scaledImage = thumbnail_img(i, 127, 127)

``thumbnail_img_noaspect``
--------------------------

Create a thumbnail of an image, without maintaining the aspect ratio.

:Synopsis: ``thumbnail_img(i, x, y, method=Image.ANTIALIAS)``

:Description: The ``thumbnail_img_noaspect`` utility creates a new image,
          scaled down so the width will be ``x``, and height will be ``y``.

:Arguments: 
 
  ``i``:
    The image to resize.

  ``x``: 
    The width.

  ``y``:
    The height.

  ``method``
    The scaling method.

:Returns: A new image. The width be ``x``, and height will be ``y``.

``thumbnail_img_square``
------------------------

Create a square thumbnail image.

:Synopsis: ``thumbnail_img_square(i, size, method=Image.ANTIALIAS)``

:Description: The ``thumbnail_img_square`` method creates a square version
              of the original image.

              * First, it scales the image so the shortest axis is ``size``
                pixels, leaving the long axis unconstrained.

              * Second, it crops the image, so the long axis is ``size``
                pixels. The cropping is done from the top-left
                corner. (There may need to be a top-right version used when
                non-Roman scripts are introduced to GroupServer.)

:Arguments:

  ``i``:
    The image to resize.

  ``size``:
    The width and height of the new image, in pixels.

  ``method``
    The scaling method.

:Returns: A new image. The width will be ``size``, and height will be
          ``size``.

Authors
=======

Thanks to Kevin for the original code:
<http://mail.python.org/pipermail/image-sig/2006-January/003724.html>.

Resources
=========

- Code repository: https://github.com/groupserver/gs.image
- Questions and comments to http://groupserver.org/groups/development
- Report bugs at https://redmine.iopen.net/projects/groupserver

.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net/
.. _Michael JasonSmith: http://groupserver.org/p/mpj17/

.. [#profiles] See ``gs.profile.image.base`` for the profile-image code:
               <https://github.com/groupserver/gs.profile.image.base/>
.. [#posts] See ``gs.group.messages.post`` for the post-rendering code:
            <https://github.com/groupserver/gs.group.messages.post/>
.. [#image] See ``gs.group.messages.image`` for the Image page:
            <https://github.com/groupserver/gs.group.messages.image/>

.. _PIL: http://www.pythonware.com/library/pil/handbook/

.. [#command] The ``get_resized`` method is a good example of why
              command-coupling is a Bad Thing.

..  LocalWords:  resizes GSSquareImage getSize px resized ANTIALIAS mpj
..  LocalWords:  retval noaspect LocalWords resize


docs/HISTORY.rst
Changelog
=========

3.1.1 (2016-09-30)
------------------

* Ensuring scaled images are always at least 1 pixel wide and 1
  pixel high, with thanks to Donald Winship for reporting `the
  issue with scaling`_

.. _the issue with scaling:
   http://groupserver.org/r/topic/1aK6lIYi2YaN6WLmAHM3nL

3.1.0 (2016-04-11)
------------------

* Making :class:`gs.image.GSImage.data` accept either strings or files
* Adding more unit tests

3.0.4 (2015-12-14)
------------------

* Ensuring an ``IOError`` when reading the cache is handled
* Updating all the unit tests, switching from a ``doctest``
* Updating the doc-strings
* Pointing at GitHub_ as the main repository

.. _GitHub: https://github.com/groupserver/gs.image

3.0.3 (2014-09-17)
------------------

* Naming all the ReStructuredText files as such

3.0.2 (2014-01-23)
------------------

* Switching to ``absolute_import``

3.0.1 (2013-09-19)
------------------

* Allowing large progressive images to scale

3.0.0 (2013-03-26)
------------------

* Added support for square images

2.0.2 (2012-12-06)
------------------

* Cleaning up the code (PEP-8)

2.0.1 (2012-10-17)
------------------

* Switching from ``PIL`` to `Pillow`_ as the main dependency

.. _Pillow: https://pillow.readthedocs.io/

2.0.0 (2011-02-15)
------------------

* Moving the Image page to `gs.group.messages.image`_

.. _gs.group.messages.image:
   https://github.com/groupserver/gs.group.messages.image

1.3.3 (2010-12-14)
------------------

* Updating the Image page so it uses arrows in the navigation
  links, and better formatting of the file-name

1.3.2 (2010-08-05)
------------------

* Fixing the ``GSImage.get_resized`` method
* Fixing the unit tests

1.3.1 (2010-07-09)
------------------

* Fixing the traversal code

1.3.0 (2009-10-20)
------------------

* Adding support for only getting the name of a cached file
* Fixing the metadata

1.2.1 (2009-09-28)
------------------

* Updating some wording
* Updating the tests

1.2.0 (2009-05-29)
------------------

* Adding a method for getting the name of the cached-file off
  disc

1.1.4 (2009-05-15)
------------------

* Dropping the Zope Image super-class, as it is persistent
* Fixing the MD5 calculation

1.1.3 (2009-05-01)
------------------

* Ensuring that ``text/html`` is used for the page-templates

1.1.2 (2008-10-02)
------------------

* Switching to anti-aliasing

1.1.1 (2008-09-26)
------------------

* Scaling the images *after* converting them to RGBA mode first

1.1.0 (2008-09-04)
------------------

* Moving code from ``GSImage`` to this product
* Using the Zope 3 image code for scaling

1.0.0 (2008-08-29)
------------------

Initial version.

..  LocalWords:  Changelog


gs/image/image.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014, 2015, 2016 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
import md5
from io import BytesIO
from logging import getLogger
log = getLogger('gs.image')
import os
from zope.app.file.image import getImageInfo
from PIL import Image as PILImage, ImageFile
from Products.XWFCore.XWFUtils import locateDataDirectory
from .utils import thumbnail_img, thumbnail_img_noaspect


class GSImage(object):

    def __init__(self, data):
        self.data = data
        self.fromCache = False

        self.data_dir = locateDataDirectory("groupserver.GSImage.cache")
        self.md5sum = md5.new(BytesIO(self.data).read()).hexdigest()
        self.base_path = os.path.join(self.data_dir, self.md5sum)

    @property
    def contentType(self):
        return self._contentType

    def _getData(self):
        return self._data

    def _setData(self, data):
        '''Set the image data.

        Nicked from the Zope 3 Image and File classes.

        ARGUMENTS
          data:   A string containing the data to set.

        RETURNS
          None.

        SIDE EFFECTS
          Sets "self._data" to contain the data.
          Sets "self._size" to the length of the data.
          Sets "self._contentType" to the content type of the image.
          Sets "self._width" to the width of the image.
          Sets "self._height" to the height of the image.'''
        if type(data) == str:
            _d = data
        elif type(data) == file:
            _d = data.read()
        else:
            m =  'Data is a {0} not a string or file'.format(type(data))
            raise TypeError(m)
        assert isinstance(_d, str)

        self._data = _d
        self._size = len(_d)
        self._contentType, self._width, self._height = getImageInfo(_d)
        self._pilImage = None

        assert isinstance(self._width, int)
        assert isinstance(self._height, int)

    data = property(_getData, _setData)

    def getSize(self):
        assert isinstance(self._size, int)
        return self._size

    def getImageSize(self):
        assert isinstance(self._width, int)
        assert isinstance(self._height, int)
        return (self._width, self._height)

    @property
    def width(self):
        assert isinstance(self._width, int)
        return self._width

    @property
    def height(self):
        assert isinstance(self._height, int)
        return self._height

    def pilImage(self):
        if self._pilImage is None:
            data_reader = BytesIO(self._data)
            self._pilImage = PILImage.open(data_reader)
        return self._pilImage

    def get_resized(self, x, y, maintain_aspect=True, only_smaller=True,
                    return_cache_path=False):
        """ Get the resized image, or the path to the resized image.

        """
        retval = None
        cache_name = self.get_cache_name(x, y, maintain_aspect,
                                         only_smaller)
        if return_cache_path:
            # possibly None if no resizing was done
            retval = cache_name
        elif cache_name:
            retval = GSImage(file(cache_name, 'rb'))
        else:
            retval = self

        return retval

    def get_cache_name(self, x, y, maintain_aspect=True, only_smaller=True):
        """SIDE EFFECTS
            Caches the image data if it scales the image
        """
        cache_name = self.base_path + '%sx%sx%s' % (x, y, maintain_aspect)
        if (only_smaller and (self._height <= y) and (self._width <= x)):
            retval = None
        elif os.path.isfile(cache_name):
            retval = cache_name
        else:
            try:
                img = self._get_resized_img(x, y, maintain_aspect)
            except IOError as e:
                m = 'Failed to get the resized image for "{0}"'
                msg = m.format(self.base_path)
                log.error(msg)
                log.error(e)
                retval = None
            else:
                self.save_img_to_cache(img, cache_name)
                retval = cache_name
        return retval

    def _get_resized_img(self, x, y, maintain_aspect):
        i = self.pilImage()
        if maintain_aspect:
            retval = thumbnail_img(i, x, y)
        else:
            retval = thumbnail_img_noaspect(i, x, y)
        return retval

    def save_img_to_cache(self, img, cache_name):
        imgFormat = self.pilImage().format
        if imgFormat == 'JPEG':
            # See Ticket 663 <https://redmine.iopen.net/issues/663>
            m = min(img.size)
            if m <= 50:
                quality = 40  # You got 40px? No quality for you.
                progressive = False
            elif m <= 200:
                quality = 60  # Medium quality.
                progressive = False
            else:  # > 200
                # This is *high* according to the JPEG standard.
                quality = 75
                progressive = True
            try:
                img.save(cache_name, imgFormat, quality=quality,
                         progressive=progressive, optimize=True)
            except IOError:
                # --=mj17=-- Increase the size of the "block" used by
                # libjpeg so we can scale stupidly large images, and retain
                # progressive rendering.
                # http://stackoverflow.com/questions/6788398/
                s = img.size[0] * img.size[1]
                ImageFile.MAXBLOCK = s
                img.save(cache_name, imgFormat, quality=quality,
                         progressive=progressive, optimize=True)
        else:
            img.save(cache_name, imgFormat)

    def _clean_cache(self):
        """ Tidy up files that have been saved in association with this
        image."""
        for fname in os.listdir(self.data_dir):
            if fname.find(self.md5sum) == 0:
                os.remove(os.path.join(self.data_dir, fname))
